google/cloud/aiplatform/datasets/tabular_dataset.py

```diff
@@ -33,6 +39,151 @@ class TabularDataset(datasets._Dataset):
+
+        input_config = metadata.get("inputConfig")
+
+        if input_config is None:
+            raise RuntimeError("No inputConfig found for dataset")
+
+        gcs_source = input_config.get("gcsSource")
+        bq_source = input_config.get("bigquerySource")
+
+        if gcs_source:
+            gcs_source_uris = gcs_source.get("uri")
+
+            if gcs_source_uris and len(gcs_source_uris) > 0:
+                # Lexicographically sort the files
+                gcs_source_uris.sort()
+
+                # Get the first file in sorted list
+                return TabularDataset._retrieve_gcs_source_columns(
+                    self.project, gcs_source_uris[0]
+                )
+        elif bq_source:
+            bq_table_uri = bq_source.get("uri")
+            if bq_table_uri:
+                return TabularDataset._retrieve_bq_source_columns(
+                    self.project, bq_table_uri
+                )
+
+        raise RuntimeError("No valid CSV or BigQuery datasource found.")
+
+    @classmethod
+    def _retrieve_gcs_source_columns(
+        cls, project: str, gcs_csv_file_path: str
+    ) -> List[str]:
+        """Retrieve the columns from a comma-delimited CSV file stored on Google Cloud Storage
+
+        Example Usage:
+
+            column_names = _retrieve_gcs_source_columns(
+                "project_id",
+                "gs://example-bucket/path/to/csv_file"
+            )
+
+            # column_names = ["column_1", "column_2"]
+
+        Args:
+            project (str):
+                Required. Project to initiate the Google Cloud Storage client with.
+            gcs_csv_file_path (str):
+                Required. A full path to a CSV files stored on Google Cloud Storage.
+                Must include "gs://" prefix.
+
+        Returns:
+            List[str]
+                A list of columns names in the CSV file.
+                
+        Raises:
+            RuntimeError: When the retrieved CSV file is invalid.
+        """
+
+        gcs_bucket, gcs_blob = utils.extract_bucket_and_prefix_from_gcs_path(
+            gcs_csv_file_path
+        )
+        client = storage.Client(project=project)
+        bucket = client.bucket(gcs_bucket)
+        blob = bucket.blob(gcs_blob)
+
+        # Incrementally download the CSV file until the header is retrieved
+        first_new_line_index = -1
+        start_index = 0
+        increment = 1000
+        line = ""
+
+        try:
+            while first_new_line_index == -1:
+                line += blob.download_as_bytes(
+                    start=start_index, end=start_index + increment
+                ).decode("utf-8")
+                first_new_line_index = line.find("\n")
+                start_index += increment
+
+            header_line = line[:first_new_line_index]
+
+            # Split to make it an iterable
+            header_line = header_line.split("\n")
+
+            csv_reader = csv.reader(header_line, delimiter=",")
```

Comment:
Leverage  CSV to read the csv format

Reply:
Hm it's a little hard to define "compatible".

One way to define compatible is to make sure it matches the exact behaviour as the service code. That would be hard to keep in sync as it can change over time.

I think it's best to add a note in the docstring that only CSV's formatted in the default ("excel") dialect are supported.

@sasha-gitg what do you think?

Reply:
Here are the docs in any case: https://cloud.google.com/ai-platform-unified/docs/datasets/prepare-tabular?hl=en#csv

My thinking is that, the user wouldn't be able to make a dataset in the first place if their data was in the wrong format (it should error out at dataset creation, but not sure).

So we can assume that, when they call `column_names` their data is already comma-delimited.
That's why I used `csv.reader(header_line, delimiter=",")`

Reply:
SGTM. No need to add the dialect to the docstring.